File: swift/leet_code/1289.swift
/**
# LeetCode 1289 Solution
This module provides a solution for the LeetCode problem 1289, "Minimum Falling Path Sum II". The problem is about finding the minimum sum of a falling path through a given grid. A falling path starts at any element in the first row and chooses the element in the next row which is not in the same column as the current element.

The `Solution` class contains two methods:
- `getTwoMin(_:)` which returns the two smallest numbers in a given row and the index of the smallest number.
- `minFallingPathSum(_:)` which calculates the minimum falling path sum in a given grid.

This module does not handle any errors as it assumes that the input is always a valid grid.

- Note: This module does not modify the input grid.
*/

class Solution {
    /**
    This method returns the two smallest numbers in a given row and the index of the smallest number.

    - Parameters:
       - row: An array of integers representing a row in the grid.

    - Returns: A tuple containing the smallest number, its index, and the second smallest number in the row.

    - Complexity: This method runs in O(n) time where n is the number of elements in the row.
    */
    private func getTwoMin(_ row: [Int]) -> (Int, Int, Int) {
        var (min1, except, min2) = (Int.max, 0, Int.max)

        for (num, i) in row.enumerated() {
            if i < min1 {
                min2 = min1
                min1 = i
                except = num
            } else if i < min2 {
                min2 = i
            }
        }

        return (min1, except, min2)
    }

    /**
    This method calculates the minimum falling path sum in a given grid.

    - Parameters:
       - grid: A 2D array of integers representing the grid.

    - Returns: The minimum falling path sum as an integer.

    - Complexity: This method runs in O(n^2) time where n is the number of rows in the grid.
    */
    func minFallingPathSum(_ grid: [[Int]]) -> Int {
        var resulting: [[Int]] = grid
        var n = grid.count

        for j in 0 ..< n - 1 {
            let (min1, except, min2) = getTwoMin(resulting[j])

            for i in 0 ..< n {
                resulting[j + 1][i] += except == i ? min2 : min1
            }
        }

        return resulting.last!.min()!
    }
}

File: dart/leet_code/1289.dart
/// Module: leet_code/1289.dart
/// This module contains a solution for finding the minimum falling path sum in
/// a grid. The grid is a 2D list of integers. The falling path starts from the
/// top row and chooses one element from each row. The next row's choice must be
/// in a column that is different from the previous row's column by at most one.
///
/// Class: Solution
/// This class encapsulates the solution for the problem.
import 'dart:math';

class Solution {
  /// Constant: MAXINT
  /// This constant represents the maximum possible integer value.
  static const MAXINT = 0x7FFFFFFFFFFFFFFF;

  /// Method: minFallingPathSum
  /// This method calculates the minimum falling path sum in the given grid.
  ///
  /// Parameters:
  /// grid - A 2D list of integers representing the grid.
  ///
  /// Returns:
  /// An integer representing the minimum falling path sum.
  ///
  /// Error Handling:
  /// This method does not explicitly handle errors. It assumes that the input
  /// grid is a non-empty 2D list of integers.
  int minFallingPathSum(List<List<int>> grid) {
    // Initialize the first and second minimum values and their indices for the first row.
    var FirstMin1 = 0, secondMin1 = 0, index1 = -1;

    // Iterate over each row in the grid.
    for (var row in grid) {
      // Initialize the first and second minimum values and their indices for the current row.
      var firstMin2 = MAXINT, secondMin2 = MAXINT, index2 = -1;

      // Iterate over each cell in the current row.
      for (var (currIndex, cell) in row.indexed) {
        // Calculate the current minimum value.
        var currMin = currIndex == index1 ? secondMin1 : FirstMin1;

        // Update the first and second minimum values and their indices for the current row.
        if (cell + currMin < firstMin2) {
          (secondMin2, firstMin2, index2) = (firstMin2, cell + currMin, currIndex);
        } else {
          secondMin2 = min(secondMin2, cell + currMin);
        }
      }
      // Update the first and second minimum values and their indices for the next row.
      (FirstMin1, secondMin1, index1) = (firstMin2, secondMin2, index2);
    }
    // Return the minimum falling path sum.
    return FirstMin1;
  }
}
